File: backend/metrics.py
"""
metrics.py — Real fairness metric computation
All values derived from actual model predictions and ground truth labels.
No synthetic/hardcoded numbers.
"""
import numpy as np


def demographic_parity_difference(y_pred: np.ndarray, groups: np.ndarray) -> dict:
    """
    DPD = P(Ŷ=1 | A=a) - P(Ŷ=1 | A=b)
    Measures if positive prediction rate differs across groups.
    Ideal = 0. Negative means group A is under-selected.
    """
    unique = np.unique(groups)
    rates = {}
    for g in unique:
        mask = groups == g
        if mask.sum() == 0:
            continue
        rates[g] = float(y_pred[mask].mean())

    if len(rates) < 2:
        return {"error": "Need at least 2 groups", "rates": rates}

    vals = list(rates.values())
    max_rate = max(vals)
    min_rate = min(vals)
    max_grp = [g for g, r in rates.items() if r == max_rate][0]
    min_grp = [g for g, r in rates.items() if r == min_rate][0]

    return {
        "rates": {k: round(v, 4) for k, v in rates.items()},
        "dpd": round(min_rate - max_rate, 4),
        "max_group": str(max_grp),
        "min_group": str(min_grp),
        "max_rate": round(max_rate, 4),
        "min_rate": round(min_rate, 4),
        "passes_threshold": abs(min_rate - max_rate) <= 0.05,
    }


def disparate_impact_ratio(y_pred: np.ndarray, groups: np.ndarray) -> dict:
    """
    DIR = P(Ŷ=1 | A=min) / P(Ŷ=1 | A=max)
    EEOC 80% rule: DIR must be >= 0.8 to pass.
    """
    unique = np.unique(groups)
    rates = {}
    for g in unique:
        mask = groups == g
        if mask.sum() == 0:
            continue
        rates[g] = float(y_pred[mask].mean())

    if len(rates) < 2:
        return {"error": "Need at least 2 groups"}

    vals = list(rates.values())
    max_rate = max(vals)
    min_rate = min(vals)
    dir_val = round(min_rate / max_rate, 4) if max_rate > 0 else 1.0

    return {
        "dir": dir_val,
        "passes_80_rule": dir_val >= 0.80,
        "legal_threshold": 0.80,
        "rates": {k: round(v, 4) for k, v in rates.items()},
    }


def equal_opportunity_difference(
    y_true: np.ndarray, y_pred: np.ndarray, groups: np.ndarray
) -> dict:
    """
    EOD = TPR(A=min) - TPR(A=max)
    True Positive Rate gap: are qualified candidates from all groups
    equally likely to be correctly selected?
    """
    unique = np.unique(groups)
    tprs = {}
    for g in unique:
        mask = groups == g
        yt = y_true[mask]
        yp = y_pred[mask]
        positives = yt == 1
        if positives.sum() == 0:
            continue
        tprs[g] = float(yp[positives].mean())

    if len(tprs) < 2:
        return {"error": "Not enough positive examples per group", "tprs": tprs}

    vals = list(tprs.values())
    max_tpr = max(vals)
    min_tpr = min(vals)

    return {
        "tprs": {k: round(v, 4) for k, v in tprs.items()},
        "eod": round(min_tpr - max_tpr, 4),
        "max_tpr": round(max_tpr, 4),
        "min_tpr": round(min_tpr, 4),
        "passes_threshold": abs(min_tpr - max_tpr) <= 0.05,
    }


def false_positive_rate_gap(
    y_true: np.ndarray, y_pred: np.ndarray, groups: np.ndarray
) -> dict:
    """
    FPR gap = FPR(A=max) - FPR(A=min)
    Are some groups falsely approved more often than others?
    """
    unique = np.unique(groups)
    fprs = {}
    for g in unique:
        mask = groups == g
        yt = y_true[mask]
        yp = y_pred[mask]
        negatives = yt == 0
        if negatives.sum() == 0:
            continue
        fprs[g] = float(yp[negatives].mean())

    if len(fprs) < 2:
        return {"error": "Not enough negative examples per group"}

    vals = list(fprs.values())
    max_fpr = max(vals)
    min_fpr = min(vals)

    return {
        "fprs": {k: round(v, 4) for k, v in fprs.items()},
        "fpr_gap": round(max_fpr - min_fpr, 4),
        "max_fpr": round(max_fpr, 4),
        "min_fpr": round(min_fpr, 4),
        "passes_threshold": abs(max_fpr - min_fpr) <= 0.05,
    }


def predictive_parity(
    y_true: np.ndarray, y_pred: np.ndarray, groups: np.ndarray
) -> dict:
    """
    Precision gap: P(Y=1 | Ŷ=1, A=a) vs P(Y=1 | Ŷ=1, A=b)
    When the model says 'hire', is it equally right across groups?
    """
    unique = np.unique(groups)
    precisions = {}
    for g in unique:
        mask = groups == g
        yt = y_true[mask]
        yp = y_pred[mask]
        predicted_pos = yp == 1
        if predicted_pos.sum() == 0:
            continue
        precisions[g] = float(yt[predicted_pos].mean())

    if len(precisions) < 2:
        return {"error": "Not enough predicted positives per group"}

    vals = list(precisions.values())
    return {
        "precisions": {k: round(v, 4) for k, v in precisions.items()},
        "pp_diff": round(min(vals) - max(vals), 4),
        "passes_threshold": abs(min(vals) - max(vals)) <= 0.05,
    }


def calibration_by_group(
    y_true: np.ndarray, y_prob: np.ndarray, groups: np.ndarray, bins: int = 5
) -> dict:
    """
    Check if predicted probabilities match actual outcomes equally across groups.
    Well-calibrated model: predicted 70% → actually hired ~70% of the time.
    """
    unique = np.unique(groups)
    calibration = {}
    bin_edges = np.linspace(0, 1, bins + 1)

    for g in unique:
        mask = groups == g
        yt = y_true[mask]
        yp = y_prob[mask]
        bin_data = []
        for i in range(bins):
            lo, hi = bin_edges[i], bin_edges[i + 1]
            in_bin = (yp >= lo) & (yp < hi)
            if in_bin.sum() < 3:
                continue
            bin_data.append({
                "range": f"{lo:.1f}-{hi:.1f}",
                "predicted_mean": round(float(yp[in_bin].mean()), 3),
                "actual_rate": round(float(yt[in_bin].mean()), 3),
                "n": int(in_bin.sum()),
            })
        calibration[g] = bin_data

    return {"calibration": calibration}


def score_distribution_stats(y_prob: np.ndarray, groups: np.ndarray) -> dict:
    """
    Distribution of predicted scores per group.
    Reveals if model systematically scores one group lower.
    """
    unique = np.unique(groups)
    stats = {}
    for g in unique:
        mask = groups == g
        scores = y_prob[mask]
        if len(scores) == 0:
            continue
        stats[g] = {
            "mean":   round(float(scores.mean()), 4),
            "median": round(float(np.median(scores)), 4),
            "std":    round(float(scores.std()), 4),
            "p25":    round(float(np.percentile(scores, 25)), 4),
            "p75":    round(float(np.percentile(scores, 75)), 4),
            "n":      int(len(scores)),
        }
    return {"score_distributions": stats}


def compute_all_metrics(
    y_true: np.ndarray,
    y_pred: np.ndarray,
    y_prob: np.ndarray,
    groups: np.ndarray,
    attr_name: str,
) -> dict:
    """Run all metrics for a given sensitive attribute. Returns a clean dict."""
    return {
        "attribute": attr_name,
        "demographic_parity": demographic_parity_difference(y_pred, groups),
        "disparate_impact": disparate_impact_ratio(y_pred, groups),
        "equal_opportunity": equal_opportunity_difference(y_true, y_pred, groups),
        "fpr_gap": false_positive_rate_gap(y_true, y_pred, groups),
        "predictive_parity": predictive_parity(y_true, y_pred, groups),
        "score_distributions": score_distribution_stats(y_prob, groups),
        "calibration": calibration_by_group(y_true, y_prob, groups),
    }

File: frontend/public/src/metrics.js
/**
 * metrics.js — Model Fairness + Feature/Proxy Analysis panels.
 * All metric values come from API.fairness() which calls the real model.
 */

const MetricsPanel = {
  _currentAttr: "gender",
  _cache: {},

  render() {
    return `
    <div class="panel" id="panel-model">
      <div class="page-hd">
        <div class="page-title">Model Fairness Metrics</div>
        <div class="page-sub">
          All metrics computed from a real GradientBoosting model trained on your data.
          Values update when you switch dimensions.
        </div>
      </div>
      <div style="display:flex;align-items:center;gap:12px;margin-bottom:22px;">
        <span style="font-size:12px;color:var(--t3)">Sensitive attribute:</span>
        <select class="flt-select" id="dim-select">
          <option value="gender">Gender</option>
          <option value="race_ethnicity">Race / Ethnicity</option>
          <option value="age_band">Age Band</option>
          <option value="disability_status">Disability Status</option>
          <option value="region">Region</option>
          <option value="education_tier">Education Tier</option>
          <option value="employment_gap">Employment Gap</option>
          <option value="nationality">Nationality</option>
        </select>
        <div class="spinner" id="metrics-spinner" style="display:none"></div>
      </div>
      <div id="metrics-banner"></div>
      <div class="metric-grid" id="metric-boxes"></div>
      <div class="cbox">
        <div class="cbox-title" id="perf-title">HIRE RATE BY GROUP — YOUR DATA</div>
        <div class="ch" style="height:250px"><canvas id="c-perf"></canvas></div>
        <div class="legend" id="perf-legend"></div>
      </div>
      <div class="cbox-row">
        <div class="cbox">
          <div class="cbox-title" id="score-avg-title">AVERAGE SCORES BY GROUP</div>
          <div class="ch" style="height:200px"><canvas id="c-score-avg"></canvas></div>
        </div>
        <div class="cbox">
          <div class="cbox-title">HIRE RATE BY GROUP — SUMMARY</div>
          <div id="group-rate-table" style="padding-top:8px;overflow-y:auto;max-height:200px"></div>
        </div>
      </div>
      <div class="cbox">
        <div class="cbox-title" id="calib-title">CALIBRATION — PREDICTED VS ACTUAL HIRE RATE</div>
        <div class="ch" style="height:240px"><canvas id="c-calib"></canvas></div>
      </div>
    </div>

    <div class="panel" id="panel-features">
      <div class="page-hd">
        <div class="page-title">Feature & Proxy Analysis</div>
        <div class="page-sub">
          Feature importance from the trained model + Cramér's V proxy scores computed from your data.
        </div>
      </div>
      <div class="sec">
        <div class="sec-title">Feature Importance (from GradientBoosting)</div>
        <div class="sec-sub">How much each numeric feature contributes to the model's predictions — computed from real trained model weights</div>
        <div id="feat-importance-bars"></div>
      </div>
      <div class="sec">
        <div class="sec-title">Proxy Feature Scores (Cramér's V)</div>
        <div class="sec-sub">Correlation between non-sensitive features and protected attributes — higher = higher proxy risk</div>
        <div id="feat-proxy-bars"></div>
      </div>
      <div class="sec">
        <div class="sec-title">Cross-Distribution Charts</div>
        <div class="sec-sub">Are qualifications distributed differently across gender groups?</div>
        <div class="cbox-row">
          <div class="cbox"><div class="cbox-title">RESUME SCORE BY GENDER</div><div class="ch" style="height:210px"><canvas id="c-feat-resume"></canvas></div></div>
          <div class="cbox"><div class="cbox-title">SKILL MATCH BY GENDER</div><div class="ch" style="height:210px"><canvas id="c-feat-skill"></canvas></div></div>
        </div>
      </div>
    </div>`;
  },

  async load() {
    const sel = document.getElementById("dim-select");
    if (sel) sel.addEventListener("change", () => this.loadAttr(sel.value));
    await this.loadAttr("gender");
  },

  async loadAttr(attr) {
    this._currentAttr = attr;
    if (this._cache[attr]) {
      this._render(attr, this._cache[attr]);
      return;
    }
    const sp = document.getElementById("metrics-spinner");
    if (sp) sp.style.display = "inline-block";
    document.getElementById("metric-boxes").innerHTML =
      `<div class="loading-row" style="grid-column:span 2"><div class="spinner"></div> Computing fairness metrics for ${attr}…</div>`;

    try {
      const data = await API.fairness(attr);
      this._cache[attr] = data;
      this._render(attr, data);
      App.log("metrics", `Fairness metrics loaded`, `Attribute: ${attr}`, "#60a5fa");
    } catch (e) {
      document.getElementById("metric-boxes").innerHTML =
        `<div class="banner high" style="grid-column:span 2"><span class="blabel">ERROR</span><span class="btext">${e.message}</span></div>`;
    } finally {
      if (sp) sp.style.display = "none";
    }
  },

  _render(attr, data) {
    const dp  = data.demographic_parity || {};
    const di  = data.disparate_impact   || {};
    const eo  = data.equal_opportunity  || {};
    const fpr = data.fpr_gap            || {};
    const sd  = data.score_distributions?.score_distributions || {};
    const cal = data.calibration?.calibration || {};

    // Banner
    const passes = di.passes_80_rule;
    document.getElementById("metrics-banner").innerHTML = `
      <div class="banner ${passes ? "low" : "high"}">
        <span class="blabel">${passes ? "PASSES 80% RULE" : "FAILS 80% RULE"}</span>
        <span class="btext">
          Disparate Impact Ratio: <strong>${di.dir ?? "N/A"}</strong>
          (${passes ? "≥" : "<"} 0.80 threshold) ·
          DPD: <strong>${dp.dpd ?? "N/A"}</strong> ·
          Worst group: <strong>${dp.min_group ?? "N/A"}</strong>
          (${((dp.min_rate ?? 0)*100).toFixed(1)}% hire rate) vs
          best: <strong>${dp.max_group ?? "N/A"}</strong>
          (${((dp.max_rate ?? 0)*100).toFixed(1)}%)
        </span>
      </div>`;

    // Metric boxes
    const mkBox = (name, val, threshold, explain, lowerBetter = false) => {
      const n = parseFloat(val);
      let cls = "ok";
      if (!isNaN(n)) {
        if (lowerBetter) cls = Math.abs(n) > Math.abs(threshold) ? (Math.abs(n) > Math.abs(threshold)*1.5 ? "danger" : "warn") : "ok";
        else cls = n < threshold ? (n < threshold * 1.5 ? "danger" : "warn") : "ok";
      }
      const pct = !isNaN(n) ? Math.min(90, Math.max(5, (n + 0.5) / 1.0 * 100)) : 50;
      return `
        <div class="mbox">
          <div class="mname">${name}</div>
          <div class="mval ${cls}">${val ?? "N/A"}</div>
          <div class="mexplain">${explain}</div>
          <div class="mbar"><div class="mbar-dot" style="left:${pct}%"></div></div>
        </div>`;
    };

    document.getElementById("metric-boxes").innerHTML =
      mkBox("Demographic Parity Difference", dp.dpd, -0.05,
        `P(Ŷ=1|A=${dp.min_group}) − P(Ŷ=1|A=${dp.max_group}). Ideal = 0. Threshold: −0.05. ${dp.passes_threshold ? "✓ Passes" : "✗ Fails"}.`, true) +
      mkBox("Disparate Impact Ratio", di.dir, 0.80,
        `P(Ŷ=1|min group) / P(Ŷ=1|max group). Legal threshold: 0.80 (EEOC). ${di.passes_80_rule ? "✓ Passes" : "✗ Fails"} 80% rule.`) +
      mkBox("Equal Opportunity Difference", eo.eod, -0.05,
        `TPR(min) − TPR(max). Are qualified applicants from all groups equally likely to be correctly selected? ${eo.passes_threshold ? "✓ Passes" : "✗ Fails"}.`, true) +
      mkBox("FPR Gap", fpr.fpr_gap, 0.05,
        `FPR(max) − FPR(min). Are unqualified applicants from some groups approved more often? ${fpr.passes_threshold ? "✓ OK" : "✗ Gap detected"}.`, true);

    // Hire rate chart — from real model predictions
    document.getElementById("perf-title").textContent =
      `${attr.toUpperCase().replace(/_/g," ")} — HIRE RATE (YOUR DATA)`;
    const rates = dp.rates || {};
    if (Object.keys(rates).length) {
      const groups = Object.entries(rates).map(([label, rate]) => ({ label, rate }));
      Charts.hireRateBar("c-perf", groups, { rotate: 20 });
      document.getElementById("perf-legend").innerHTML =
        groups.map((g, i) =>
          `<span><span class="lsw" style="background:${Charts.CHART_COLORS[i]}"></span>${g.label} (${(g.rate*100).toFixed(1)}%)</span>`
        ).join("");
    }

    // Score distributions — fetch stats and build charts
    this._renderScoreCharts(attr);

    // Calibration chart
    if (Object.keys(cal).length) {
      Charts.calibrationChart("c-calib", cal);
    }
    document.getElementById("calib-title").textContent =
      `CALIBRATION — ${attr.toUpperCase().replace(/_/g," ")} (PREDICTED PROBABILITY VS ACTUAL OUTCOME RATE)`;
  },

  async _renderScoreCharts(attr) {
    try {
      const stats      = await API.stats();
      const groupStats = stats.group_stats?.[attr] || {};
      const groups     = Object.keys(groupStats);
      if (!groups.length) return;

      // Chart 1: Average numeric scores by group (use hire_rate as proxy)
      const numByGroup = stats.numeric_by_group || {};
      const scoreKey   = Object.keys(numByGroup).find(k =>
        k.includes("score") && k.includes(`_by_${attr}`));

      document.getElementById("score-avg-title").textContent =
        scoreKey
          ? `AVG ${scoreKey.replace(`_by_${attr}`,"").replace(/_/g," ").toUpperCase()} BY ${attr.replace(/_/g," ").toUpperCase()}`
          : `HIRE RATE BY ${attr.replace(/_/g," ").toUpperCase()}`;

      const chartData = scoreKey
        ? { labels: Object.keys(numByGroup[scoreKey]),
            values: Object.values(numByGroup[scoreKey]).map(v => parseFloat(v.toFixed(3))) }
        : { labels: groups,
            values: groups.map(g => groupStats[g].hire_rate) };

      Charts.barChart("c-score-avg", chartData.labels, [{
        label: "Score",
        data:  chartData.values,
        backgroundColor: Charts.CHART_COLORS,
        borderRadius: 4,
      }], { legend: false, max: 1 });

      // Table 2: Hire rate table with counts
      const tableEl = document.getElementById("group-rate-table");
      tableEl.innerHTML = `<table class="tbl" style="font-size:12px">
        <thead><tr><th>Group</th><th>Total</th><th>Selected</th><th>Rate</th></tr></thead>
        <tbody>
        ${groups.map(g => {
          const d   = groupStats[g];
          const pct = (d.hire_rate * 100).toFixed(1);
          const cls = d.hire_rate < 0.15 ? "var(--red)" : d.hire_rate < 0.25 ? "var(--amber)" : "var(--green)";
          return `<tr>
            <td>${g}</td>
            <td style="font-family:var(--fm)">${d.total}</td>
            <td style="font-family:var(--fm)">${d.hired}</td>
            <td style="font-family:var(--fm);color:${cls}">${pct}%</td>
          </tr>`;
        }).join("")}
        </tbody></table>`;
    } catch (e) {
      console.error("Score chart error:", e);
    }
  },

  // Feature panel loader
  async loadFeatures() {
    try {
      const [trainData, proxyData] = await Promise.all([
        Promise.resolve(App.trainResult || {}),
        API.proxy(),
      ]);

      // Feature importance from real model
      const fi = trainData.feature_importance || {};
      const fiEl = document.getElementById("feat-importance-bars");
      if (Object.keys(fi).length) {
        const maxFI = Math.max(...Object.values(fi));
        fiEl.innerHTML = Object.entries(fi)
          .sort((a,b) => b[1]-a[1])
          .map(([name, val]) => `
            <div class="feat-row">
              <div class="feat-name">${name}</div>
              <div class="feat-bg"><div class="feat-fill" style="width:${(val/maxFI*100).toFixed(1)}%;background:#9d8df8"></div></div>
              <div class="feat-pct">${(val*100).toFixed(1)}%</div>
            </div>`).join("");
      } else {
        fiEl.innerHTML = `<div class="banner info"><span class="blabel">N/A</span><span class="btext">Feature importance not available — train the model first.</span></div>`;
      }

      // Proxy bars from real Cramér's V
      const risks = proxyData.feature_proxy_risks || {};
      const proxyEl = document.getElementById("feat-proxy-bars");
      if (Object.keys(risks).length) {
        const sorted = Object.entries(risks).sort((a,b) => b[1].max_proxy_score - a[1].max_proxy_score);
        proxyEl.innerHTML = sorted.map(([feat, r]) => `
          <div class="feat-row">
            <div class="feat-name">
              ${feat}
              <span class="pill ${r.risk_level==="HIGH"?"red":"amber"}" style="font-size:9px;padding:1px 4px">${r.risk_level}</span>
            </div>
            <div class="feat-bg"><div class="feat-fill" style="width:${(r.max_proxy_score*100).toFixed(1)}%;background:${r.risk_level==="HIGH"?"#f87171":"#fbbf24"}"></div></div>
            <div class="feat-pct">${r.max_proxy_score}</div>
          </div>`).join("");
      } else {
        proxyEl.innerHTML = `<div class="banner low"><span class="blabel">CLEAN</span><span class="btext">No proxy features detected above threshold.</span></div>`;
      }

      // Load score data for cross-distribution charts
      const statsData = await API.stats();
      const gs = statsData.group_stats?.gender || {};
      // Build from group hire rates as proxy for score distribution shape
      // (real histograms would need raw score bins from backend)
      ["c-feat-resume", "c-feat-skill"].forEach(cid => {
        const labels = Object.keys(gs);
        Charts.barChart(cid, labels,
          [{ label:"Hire Rate", data: labels.map(g => gs[g]?.hire_rate || 0), backgroundColor: Charts.CHART_COLORS, borderRadius:4 }],
          { pct: true });
      });

    } catch (e) {
      document.getElementById("feat-importance-bars").innerHTML =
        `<div class="banner high"><span class="blabel">ERROR</span><span class="btext">${e.message}</span></div>`;
    }
  },
};

window.MetricsPanel = MetricsPanel;